discoverd2/server/state.go

```diff
@@ -199,93 +196,185 @@ type State struct {
 	subscribersMtx sync.Mutex
 }
 
-func (s *State) AddService(service string) {
+func newService() *service {
+	return &service{
+		instances: make(map[string]*Instance),
+	}
+}
+
+type service struct {
+	// instance ID -> instance
+	instances map[string]*Instance
+
+	leaderID     string
+	leaderIndex  uint64
```

Comment:
Perhaps a quick comment that this is usually constrained to being `>0` as per the `Instance.Index` field, with a sentinel value for zero -> unset?